core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -770,9 +794,19 @@ private[spark] class TaskSetManager(
         logError("Unknown TaskEndReason: " + e)
         None
     }
-    // always add to failed executors
-    failedExecutors.getOrElseUpdate(index, new HashMap[String, Long]()).
-      put(info.executorId, clock.getTimeMillis())
+
+    // we might rack up a bunch of fetch-failures in rapid succession, due to a bad node.  But
+    // that bad node will get handled separately by spark's stage-failure handling mechanism.  It
+    // shouldn't penalize *this* executor at all, so don't count it as a task-failure as far as
+    // the blacklist is concerned.
+    val countTowardsTaskFailures = reason match {
+      case fail: TaskFailedReason => fail.countTowardsTaskFailures
```

Comment:
the code on the old line 710 makes it seem like we always expect this to be a TaskFailedReason (should the param type just be changed?)?